File: src/server/api/sync_crdt_handler_test.go
package api

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	_ "github.com/mattn/go-sqlite3"
)

func setupTestDB(t *testing.T) *sql.DB {
	db, err := sql.Open("sqlite3", ":memory:")
	if err != nil {
		t.Fatalf("Failed to open DB: %v", err)
	}

	_, err = db.Exec(`
		CREATE TABLE crdt_deltas (
			id TEXT PRIMARY KEY,
			entity_id TEXT,
			data TEXT,
			updated_at TEXT
		)
	`)
	if err != nil {
		t.Fatalf("Failed to create table: %v", err)
	}

	return db
}

func TestSyncCRDTHandler(t *testing.T) {
	db := setupTestDB(t)
	defer db.Close()

	handler := NewSyncCRDTHandler(db)

	payload := CRDTPayload{
		Deltas: []CRDTDelta{
			{ID: "1", EntityID: "e1", Data: "d1", UpdatedAt: "2023-01-01T00:00:00Z"},
		},
	}
	body, _ := json.Marshal(payload)
	req := httptest.NewRequest("POST", "/api/v1/sync/mcp-deltas", bytes.NewBuffer(body))
	w := httptest.NewRecorder()

	handler.HandlePost(w, req)

	if w.Code != http.StatusOK {
		t.Errorf("Expected status 200, got %d", w.Code)
	}

	var count int
	err := db.QueryRow("SELECT COUNT(*) FROM crdt_deltas").Scan(&count)
	if err != nil {
		t.Errorf("Failed to query DB: %v", err)
	}
	if count != 1 {
		t.Errorf("Expected 1 row, got %d", count)
	}
}

File: src/server/api/sync_crdt_handler_test.rs
use axum::{body::Body, http::{Request, StatusCode}};
use tower::ServiceExt;
use sqlx::PgPool;
use super::sync_crdt_handler::{handle_crdt_sync, CrdtSyncRequest, CrdtDelta};
use serde_json::json;

#[tokio::test]
async fn test_handle_crdt_sync() {
    let pool_opts = sqlx::postgres::PgPoolOptions::new()
        .max_connections(1);
    let pool = match pool_opts.connect("postgres://postgres:postgres@localhost:5432/test").await {
        Ok(pool) => pool,
        Err(_) => return, // Skip if db not available
    };

    // Ensure table exists for test
    sqlx::query("CREATE TABLE IF NOT EXISTS crdt_deltas (id TEXT PRIMARY KEY, entity_id TEXT NOT NULL, data TEXT NOT NULL, updated_at TIMESTAMPTZ NOT NULL)")
        .execute(&pool)
        .await
        .unwrap();

    let app = axum::Router::new()
        .route("/api/v1/sync/mcp-deltas", axum::routing::post(handle_crdt_sync))
        .with_state(pool);

    let req_body = json!({
        "deltas": [
            {
                "id": "delta_123",
                "entity_id": "task_abc",
                "data": "{\"status\": \"completed\"}",
                "updated_at": "2026-04-17T12:00:00Z"
            }
        ]
    });

    let request = Request::builder()
        .method("POST")
        .uri("/api/v1/sync/mcp-deltas")
        .header("content-type", "application/json")
        .body(Body::from(req_body.to_string()))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    assert_eq!(response.status(), StatusCode::OK);
}
